File: aoc2017/day8.swift
//
//  day8.swift
//  aoc2017
//
//  Created by Chris McElroy on 7/23/21.
//

import Foundation

func day8() {
	var reg: [String: Int] = [:]
	
	let input = inputWords(8)
	var max = 0
	
	for line in input {
		var pos = -1
		if line[1] == "inc" {
			pos = 1
		}
//		let amt = Int(line[2])!
		switch line[5] {
		case "<":
			if reg[line[4], default: 0] < Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		case ">":
			if reg[line[4], default: 0] > Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		case "<=":
			if reg[line[4], default: 0] <= Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		case ">=":
			if reg[line[4], default: 0] >= Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		case "==":
			if reg[line[4], default: 0] == Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		case "!=":
			if reg[line[4], default: 0] != Int(line[6])! {
				reg[line[0], default: 0] += pos*Int(line[2])!
			}
		default: break
		}
		let cMax = reg.values.max()!
		if cMax > max { max = cMax }
 	}
	
	print(max)
}

File: go/day8.go
package main

import "strconv"

func day8() {
	reg := map[string]int{}

	max1 := 0
	max2 := 0

	for _, line := range wordLines {
		var pos = -1
		if line[1] == "inc" {
			pos = 1
		}
		amt, _ := strconv.Atoi(line[2])
		compVal, _ := strconv.Atoi(line[6])
		switch line[5] {
		case "<":
			if reg[line[4]] < compVal {
				reg[line[0]] += pos * amt
			}
		case ">":
			if reg[line[4]] > compVal {
				reg[line[0]] += pos * amt
			}
		case "<=":
			if reg[line[4]] <= compVal {
				reg[line[0]] += pos * amt
			}
		case ">=":
			if reg[line[4]] >= compVal {
				reg[line[0]] += pos * amt
			}
		case "==":
			if reg[line[4]] == compVal {
				reg[line[0]] += pos * amt
			}
		case "!=":
			if reg[line[4]] != compVal {
				reg[line[0]] += pos * amt
			}
		}
		for _, v := range reg {
			if v > max2 {
				max2 = v
			}
		}
	}

	for _, v := range reg {
		if v > max1 {
			max1 = v
		}
	}
	println(max1, max2)
}
